File: src/mongo/query.ts
import { ObjectId } from "mongodb";
import type { MongoClient, Document, Filter, Sort } from "mongodb";
import { getTimeout } from "../lib/timeout.ts";
import { serializeDocuments, serializeDocument } from "./serialize.ts";

type FindOpts = {
  dbName: string;
  collName: string;
  filter?: Document;
  sort?: Sort;
  projection?: Document;
  limit: number;
  skip: number;
};

type FindResult = {
  database: string;
  collection: string;
  filter: Document;
  documents: Record<string, unknown>[];
  count: number;
  hasMore: boolean;
  totalMatching: number;
};

export async function findDocuments(client: MongoClient, opts: FindOpts): Promise<FindResult> {
  const timeout = getTimeout();
  const collection = client.db(opts.dbName).collection(opts.collName);
  const filter = (opts.filter ?? {}) as Filter<Document>;

  const cursor = collection.find(filter, { maxTimeMS: timeout });
  if (opts.sort) {
    cursor.sort(opts.sort);
  }
  if (opts.projection) {
    cursor.project(opts.projection);
  }
  cursor.skip(opts.skip);
  cursor.limit(opts.limit + 1);

  const rawDocs = await cursor.toArray();
  const hasMore = rawDocs.length > opts.limit;
  const docs = hasMore ? rawDocs.slice(0, opts.limit) : rawDocs;

  const isEmptyFilter = Object.keys(filter).length === 0;
  const totalMatching = isEmptyFilter
    ? await collection.estimatedDocumentCount({ maxTimeMS: timeout })
    : await collection.countDocuments(filter, { maxTimeMS: timeout });

  return {
    database: opts.dbName,
    collection: opts.collName,
    filter,
    documents: serializeDocuments(docs),
    count: docs.length,
    hasMore,
    totalMatching,
  };
}

type FindByIdOpts = {
  dbName: string;
  collName: string;
  rawId: string;
  idType?: string;
  projection?: Document;
};

export async function findById(
  client: MongoClient,
  opts: FindByIdOpts,
): Promise<Record<string, unknown> | null> {
  const timeout = getTimeout();
  const collection = client.db(opts.dbName).collection(opts.collName);
  const id = parseId(opts.rawId, opts.idType);
  const doc = await collection.findOne({ _id: id } as Filter<Document>, {
    maxTimeMS: timeout,
    projection: opts.projection,
  });
  return doc ? serializeDocument(doc) : null;
}

type CountOpts = {
  dbName: string;
  collName: string;
  filter?: Document;
};

export async function countDocuments(client: MongoClient, opts: CountOpts): Promise<number> {
  const timeout = getTimeout();
  const collection = client.db(opts.dbName).collection(opts.collName);
  const f = (opts.filter ?? {}) as Filter<Document>;
  const isEmpty = Object.keys(f).length === 0;
  return isEmpty
    ? collection.estimatedDocumentCount({ maxTimeMS: timeout })
    : collection.countDocuments(f, { maxTimeMS: timeout });
}

type DistinctOpts = {
  dbName: string;
  collName: string;
  field: string;
  filter?: Document;
};

export async function getDistinctValues(client: MongoClient, opts: DistinctOpts): Promise<unknown[]> {
  const timeout = getTimeout();
  const collection = client.db(opts.dbName).collection(opts.collName);
  const f = (opts.filter ?? {}) as Filter<Document>;
  const values = await collection.distinct(opts.field, f, { maxTimeMS: timeout });
  return values.map((v) => {
    if (v instanceof ObjectId) {
      return v.toHexString();
    }
    if (v instanceof Date) {
      return v.toISOString();
    }
    return v;
  });
}

function parseId(raw: string, type?: string): unknown {
  if (type === "objectid" || (!type && /^[0-9a-fA-F]{24}$/.test(raw))) {
    return new ObjectId(raw);
  }
  if (type === "number") {
    const n = Number(raw);
    if (!Number.isFinite(n)) {
      throw new Error(`Invalid number ID: "${raw}"`);
    }
    return n;
  }
  return raw;
}

File: internal/mongo/query.go
package mongo

import (
	"context"
	"errors"
	"fmt"
	"regexp"
	"slices"
	"strconv"
	"strings"

	"go.mongodb.org/mongo-driver/v2/bson"
	driver "go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"

	"github.com/shhac/agent-mongo/internal/serialize"
)

type FindOpts struct {
	Ref
	Filter     bson.D
	Sort       bson.D
	Projection bson.D
	Limit      int
	Skip       int
}

type FindResult struct {
	Documents     []map[string]any
	HasMore       bool
	TotalMatching int64
}

func orEmpty(filter bson.D) bson.D {
	if filter == nil {
		return bson.D{}
	}
	return filter
}

func (s *Session) FindDocuments(ctx context.Context, opts FindOpts) (FindResult, error) {
	raw, err := s.runCursor(ctx, opts.DB, findCommand(opts))
	if err != nil {
		return FindResult{}, err
	}

	raw, hasMore := keepLimit(raw, opts.Limit)

	totalMatching, err := s.CountDocuments(ctx, opts.Ref, opts.Filter)
	if err != nil {
		return FindResult{}, err
	}

	return FindResult{
		Documents:     serialize.Documents(raw),
		HasMore:       hasMore,
		TotalMatching: totalMatching,
	}, nil
}

// CountDocuments counts what a filter matches; with no filter, the collection's
// metadata count, which is instant where a full count scans.
func (s *Session) CountDocuments(ctx context.Context, ref Ref, filter bson.D) (int64, error) {
	if len(filter) == 0 {
		return s.estimatedCount(ctx, ref)
	}
	return s.collection(ref).CountDocuments(ctx, filter, commented(options.Count(), s.comment))
}

func (s *Session) estimatedCount(ctx context.Context, ref Ref) (int64, error) {
	return s.collection(ref).EstimatedDocumentCount(ctx,
		commented(options.EstimatedDocumentCount(), s.comment))
}

func (s *Session) collection(ref Ref) *driver.Collection {
	return s.Client.Database(ref.DB).Collection(ref.Collection)
}

type FindByIDOpts struct {
	Ref
	ID         any // from ParseID
	Projection bson.D
}

var objectIDPattern = regexp.MustCompile(`^[0-9a-fA-F]{24}$`)

// IDTypes are the --type values ParseID understands; "" auto-detects.
var IDTypes = []string{"objectid", "string", "number"}

// ParseID interprets a raw _id argument, auto-detecting ObjectIds by shape.
// Pure, so a malformed id or --type fails before anything connects.
func ParseID(raw, idType string) (any, error) {
	if idType != "" && !slices.Contains(IDTypes, idType) {
		return nil, fmt.Errorf("Invalid --type: %q. Valid: %s", idType, strings.Join(IDTypes, ", "))
	}
	if idType == "objectid" || (idType == "" && objectIDPattern.MatchString(raw)) {
		return bson.ObjectIDFromHex(raw)
	}
	if idType == "number" {
		n, err := strconv.ParseFloat(raw, 64)
		if err != nil {
			return nil, fmt.Errorf("Invalid number ID: %q", raw)
		}
		return n, nil
	}
	return raw, nil
}

// FindByID returns the serialized document, or nil when not found.
func (s *Session) FindByID(ctx context.Context, opts FindByIDOpts) (map[string]any, error) {
	findOpts := commented(options.FindOne(), s.comment)
	if opts.Projection != nil {
		findOpts = findOpts.SetProjection(opts.Projection)
	}
	var doc bson.D
	err := s.collection(opts.Ref).
		FindOne(ctx, bson.D{{Key: "_id", Value: opts.ID}}, findOpts).
		Decode(&doc)
	if err != nil {
		if errors.Is(err, driver.ErrNoDocuments) {
			return nil, nil
		}
		return nil, err
	}
	return serialize.Document(doc), nil
}

func (s *Session) DistinctValues(
	ctx context.Context, ref Ref, field string, filter bson.D,
) ([]any, error) {
	result := s.collection(ref).Distinct(ctx, field, orEmpty(filter),
		commented(options.Distinct(), s.comment))
	// Asked first because Decode ignores it despite documenting otherwise: a
	// failed distinct would surface as "error decoding key arr: unexpected EOF",
	// losing the timeout or auth failure an agent needs to act on.
	if err := result.Err(); err != nil {
		return nil, err
	}
	var values bson.A
	if err := result.Decode(&values); err != nil {
		return nil, err
	}
	out := make([]any, len(values))
	for i, v := range values {
		out[i] = serialize.Value(v)
	}
	return out, nil
}

// SampleDocuments returns random documents, optionally filtered first.
func (s *Session) SampleDocuments(
	ctx context.Context, ref Ref, size int, filter bson.D,
) ([]map[string]any, error) {
	raw, err := s.runCursor(ctx, ref.DB, aggregateCommand(ref.Collection, samplePipeline(filter, size), size))
	if err != nil {
		return nil, err
	}
	return serialize.Documents(raw), nil
}
